server/service.go

```diff
@@ -326,31 +346,49 @@ func (m *BoostService) handleGetHeader(w http.ResponseWriter, req *http.Request)
 			defer mu.Unlock()
 
 			// Skip if value (fee) is not greater than the current highest value
-			if result.Data != nil && responsePayload.Data.Message.Value.Cmp(&result.Data.Message.Value) < 1 {
+			if result.response.Data != nil && responsePayload.Data.Message.Value.Cmp(&result.response.Data.Message.Value) < 1 {
 				return
 			}
 
 			// Use this relay's response as mev-boost response because it's most profitable
-			*result = *responsePayload
-			log.Info("successfully got more valuable payload header")
+			log.Debug("received a better bid")
+			result.response = *responsePayload
+			result.relay = relay.String()
+			result.t = time.Now()
 		}(relay)
 	}
 
 	// Wait for all requests to complete...
 	wg.Wait()
 
-	if result.Data == nil || result.Data.Message == nil || result.Data.Message.Header == nil || result.Data.Message.Header.BlockHash == nilHash {
-		log.Info("no bids received from relay")
+	if result.response.Data == nil {
+		log.Info("no bid received")
 		w.WriteHeader(http.StatusNoContent)
 		return
 	}
 
-	m.respondOK(w, result)
+	// Log result
+	log.WithFields(logrus.Fields{
+		"blockNumber": result.response.Data.Message.Header.BlockNumber,
+		"blockHash":   result.response.Data.Message.Header.BlockHash,
+		"txRoot":      result.response.Data.Message.Header.TransactionsRoot.String(),
+		"value":       result.response.Data.Message.Value.String(),
+		"relay":       result.relay,
+	}).Info("best bid")
+
+	// Remember the bid, for future logging in case of withholding
+	bidKey := bidRespKey{slot: _slot, blockHash: result.response.Data.Message.Header.BlockHash.String()}
+	m.bidsLock.Lock()
+	m.bids[bidKey] = result
+	m.bidsLock.Unlock()
+
+	// Return the bid
+	m.respondOK(w, result.response)
```

Comment:
since more than one relay can respond with the same (slot, blockhash) it's probably better to save all relays responses